src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -2132,6 +2132,83 @@ int RGWMetadataHandlerPut_Bucket::put_post(const DoutPrefixProvider *dpp)
   return ret;
 }
 
+int update_bucket_topic_mappings(const DoutPrefixProvider* dpp,
+                                 RGWBucketCompleteInfo* orig_bci,
+                                 RGWBucketCompleteInfo* current_bci,
+                                 rgw::sal::Driver* driver) {
+  const auto decode_attrs =
+      [&](const rgw::sal::Attrs& attrs,
+          rgw_pubsub_bucket_topics& bucket_topics) -> int {
+    auto iter = attrs.find(RGW_ATTR_BUCKET_NOTIFICATION);
+    if (iter == attrs.end()) {
+      return 0;
+    }
+    try {
+      const auto& bl = iter->second;
+      auto biter = bl.cbegin();
+      bucket_topics.decode(biter);
+    } catch (buffer::error& err) {
+      return -EIO;
+    }
+    return 0;
+  };
+  rgw_pubsub_bucket_topics old_bucket_topics;
+  if (orig_bci) {
+    auto ret = decode_attrs(orig_bci->attrs, old_bucket_topics);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1)
+          << "ERROR: failed to decode OLD bucket topics for bucket: "
+          << orig_bci->info.bucket.name << dendl;
+      return ret;
+    }
+  }
+  rgw_pubsub_bucket_topics current_bucket_topics;
+  auto ret = decode_attrs(current_bci->attrs, current_bucket_topics);
+  if (ret < 0) {
+    ldpp_dout(dpp, 1)
+        << "ERROR: failed to decode current bucket topics for bucket: "
+        << current_bci->info.bucket.name << dendl;
+    return ret;
+  }
+  // fetch the list of subscribed topics stored inside old_bucket attrs.
+  std::unordered_map<std::string, rgw_pubsub_topic> old_topics;
+  for (const auto& [_, topic_filter] : old_bucket_topics.topics) {
+    old_topics[topic_filter.topic.name] = topic_filter.topic;
+  }
+  // fetch the list of subscribed topics stored inside current_bucket attrs.
+  std::unordered_map<std::string, rgw_pubsub_topic> current_topics;
+  for (const auto& [_, topic_filter] : current_bucket_topics.topics) {
+    current_topics[topic_filter.topic.name] = topic_filter.topic;
+  }
+  auto bucket = driver->get_bucket(current_bci->info);
+  // traverse thru old topics and check if they are not in current, then delete
+  // the mapping, if present in both current and old then delete from current
+  // set as we do not need to update those mapping.
```

Comment:
maybe it would be better to use `std::set_difference()` to get the 2 groups of topics:
old - new -> topics to remove
new - old -> topics to add

Reply:
also, it is probably better to use `cls_cmpomap`  to do these operations atomically.

Reply:
> maybe it would be better to use `std::set_difference()` to get the 2 groups of topics: old - new -> topics to remove new - old -> topics to add

i thought about using the set_difference() but it did not make sense looping multiple times (set_difference loops and creates a diff set and then i have to again loop the diff_set to update the mapping). with current approach i just loop single time and perform the update at same time, while doing the diff

Reply:
> also, it is probably better to use `cls_cmpomap` to do these operations atomically.

i think `cls_cmpomap` is bit overkill here, coz the function is either going to add the mapping or remove the mapping, current s3 create_notification/delete_notification allows you to either add/remove mapping at an given time. so not sure atomicity could be problem

Reply:
you are right regarding `cls_cmpomap` - this is meant to compare the values not the keys.
but i still think we should not update the keys one by one in a loop. e.g. `omap_rm_keys()` for deleting multiple keys

Reply:
ideally this function would send a single `librados::ObjectWriteOperation` that does both `op.omap_set()` and `op.omap_rm_keys()`

and because this metadata handler is part of the rados backend, it shouldn't need to call through `driver` or allocate a `sal::Bucket` to do that

Reply:
> ideally this function would send a single `librados::ObjectWriteOperation` that does both `op.omap_set()` and `op.omap_rm_keys()`
> 
> and because this metadata handler is part of the rados backend, it shouldn't need to call through `driver` or allocate a `sal::Bucket` to do that

I have a common function inside the sal_rados.cc, so calling it through `driver`
regarding the single call for `librados::ObjectWriteOperation`, the corresponding S3 operation that adds/update/removes S3 notifications can only single operation in single S3 call so we would either add the key or remove a key at any given time and hence `op.omap_set()` and `op.omap_rm_keys()` are called separately inside the final function that does the write.

Reply:
> you are right regarding `cls_cmpomap` - this is meant to compare the values not the keys. but i still think we should not update the keys one by one in a loop. e.g. `omap_rm_keys()` for deleting multiple keys

`omap_rm_keys()` is just deleting the single key as the `bucket_name` is the key and its going to be single key. The loop is for different topics and each is a different omap object hence cannot be called for all topics in single call.